### PR TITLE
[video_player] add updateInterval

### PR DESCRIPTION
**Pull request by lukepighetti**
This PR adds an optional `updateDuration` constructor parameter which allows the consumer to set how often the video info value notifier is updated.

Adding the `updateDuration` member to `FakeVideoPlayer` should constrain regressions (removing the member)

Resolves: https://github.com/flutter/flutter/issues/90080

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[the auto-formatter]: https://github.com/flutter/plugins/blob/master/script/tool/README.md#format-code
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/video_player/video_player/CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,7 @@
 
 * Updates minimum Flutter version to 2.10.
 * Adds OS version support information to README.
+* Adds optional updateDuration parameter.
```
Comment (hellohuanlin) [inline packages/video_player/video_player/CHANGELOG.md:5]: this should be a new version

File: packages/video_player/video_player/lib/video_player.dart [hunk 2]
```diff
@@ -207,7 +207,10 @@ class VideoPlayerController extends ValueNotifier<VideoPlayerValue> {
   /// null. The [package] argument must be non-null when the asset comes from a
   /// package and null otherwise.
   VideoPlayerController.asset(this.dataSource,
-      {this.package, this.closedCaptionFile, this.videoPlayerOptions})
+      {this.package,
+      this.closedCaptionFile,
+      this.videoPlayerOptions,
+      this.updateDuration = const Duration(milliseconds: 500)})
```
Comment (hellohuanlin) [inline packages/video_player/video_player/lib/video_player.dart:213]: should this be `valueNotifierInterval` or `updateInterval` or something like that?

File: packages/video_player/video_player/lib/video_player.dart [hunk 3]
```diff
@@ -365,6 +377,7 @@ class VideoPlayerController extends ValueNotifier<VideoPlayerValue> {
 
       switch (event.eventType) {
         case VideoEventType.initialized:
+          print('initialized2');
```
Comment (hellohuanlin) [inline packages/video_player/video_player/lib/video_player.dart:380]: this should not be checked in

File: packages/video_player/video_player/lib/video_player.dart [hunk 4]
```diff
@@ -377,6 +390,7 @@ class VideoPlayerController extends ValueNotifier<VideoPlayerValue> {
           _applyPlayPause();
           break;
         case VideoEventType.completed:
+          print('completed');
```
Comment (hellohuanlin) [inline packages/video_player/video_player/lib/video_player.dart:393]: same here

**Review by hellohuanlin — CHANGES_REQUESTED** [review]

**Review by stuartmorgan — COMMENTED** [review]
I'm not convinced that this is a good idea. Under what conditions would a value that someone customized this to be better than a default value we picked?

In particular:
- If this is only for playback position, as I said in the issue I'm not sure changing this is the best option, vs. adding animation to the provided progress bar.
- If there are other use cases, do they actually vary in what interval they need? If not, then it seems like either:
  - The current default is the best option, or
  - A different default would be a better option (with some supporting perf analysis on various hardware)

**Comment by lukepighetti** [comment]
A faster value update interval is necessary when you are trying to paint UI with frame-by-frame metadata over the existing video player. It works very well at 33 ms for a 30 fps video. This update is not a breaking change, and it allows developers to adjust it as they see fit.

**Comment by stuartmorgan** [comment]
> A faster value update interval is necessary when you are trying to paint UI with frame-by-frame metadata over the existing video player.

Can you elaborate on the use case here? What metadata in `VideoPlayerValue` is changing per-frame?

> It works very well at 33 ms for a 30 fps video.

On what hardware? If this is true as a general statement, it's not clear why this is a configuration option rather than an update to the value (per my second point above).

> This update is not a breaking change

I didn't say it was. It is, however, new API, and new API requires justification.

> and it allows developers to adjust it as they see fit.

Delegating all decisions to plugin clients is not always the right solution to a problem. This warrants more discussion about why it's the best way to address the underlying need, thus my questions above.

**Comment by lukepighetti** [comment]
We are painting our own frame indexed metadata as UI, using VideoPlayerValue to get the current position duration to lookup the metadata. The current implementation limits this UI overlay to 2 FPS. This PR allows developers to tune the VideoPlayerValue update interval to suit their needs, and it's dont with a thin PR and no detrimental effects to existing consumers.

**Comment by stuartmorgan** [comment]
> We are painting our own frame indexed metadata as UI, using VideoPlayerValue to get the current position duration to lookup the metadata.

It sounds like if the controller itself did interpolation of the current playback position, in Dart, that would address your use case then?

> This PR allows developers to tune the VideoPlayerValue update interval to suit their needs.

I understand that; I responded to this same comment in my last reply.

> It's achieved with an intentionally thin PR and no detrimental effects to existing consumers.

Approaching this discussion from the standpoint that I don't understand what the PR does is really not productive. We've already established that I understand (as is obvious from reading the PR) that it's not a breaking change. That's not my concern, and I've never said that it was.

**Comment by lukepighetti** [comment]
If there were a separate stream for duration position that supported a frequency of `1000/fps` ms and it was guaranteed to be accurate through all state transitions then there would be no need for updateInterval to the best of my knowledge.

> Approaching this discussion from the standpoint that I don't understand what the PR does is really not productive.

Not sure where I insinuated that, but I apologize if that was communicated to you.

**Comment by stuartmorgan** [comment]
> If there were a separate stream for duration position that supported a frequency of `1000/fps` ms

Why do you expect a stream API at all? It's a strange API to use an event stream for changes to what is essentially a continuous value change. I would expect the event stream not to include position at all, and for position to just be a property that interested clients can request whenever they want, whereas this PR moves in the opposite direction.

> That said, the nice thing about the updateInterval PR is that it covers a wide range of needs from "the default playhead behavior feels like it updates too slow" to "we'd like to paint frame metadata in realtime and are comfortable with the costs."

Neither of these requires a streamed position at all as far as I can tell, just a position parameter that is accurate. If you want position at an arbitrary frequency that you control, and you don't care about having it be a method channel call every time, is there some reason you aren't just calling the existing `position` getter with that frequency?

> > Approaching this discussion from the standpoint that I don't understand what the PR does is really not productive.
> 
> Not sure where I insinuated that, but I apologize if that was communicated to you.

You told me twice in a row, despite it being unrelated to any of my feedback, that a PR that adds an optional parameter whose default value is the same as the value that was previously hard coded is not a breaking change. If you believed I already understand that obvious fact, I'm not sure what you were trying to convey with those comments.

**Pull request closed without merging**